lib/sender.js

```diff
@@ -303,8 +315,9 @@ class Sender {
    * @public
    */
   send(data, options, cb) {
-    const buf = toBuffer(data);
     const perMessageDeflate = this._extensions[PerMessageDeflate.extensionName];
+    const buf =
+      perMessageDeflate || typeof data !== 'string' ? toBuffer(data) : data;
```

Comment:
The `buf.length >= perMessageDeflate._threshold` check below might not work as expected when `data` is a string because it does not use the string bytes length.

---

Also, when `data` is a string and permessage-deflate is enabled, `toBuffer()` should be ideally called only when the the message will actually be compressed (`this._compress === true`).

It would be better to make this permessage-deflate agnostic. Removing the `perMessageDeflate` check here and passing the string to `perMessageDeflate.compress()` should work. The problem is that `enqueue()` and `dequeue()` currently use the buffer length to update the `_bufferedBytes` value.

Reply:
That branch is only used if `perMessageDeflate` in which case it's converted to a buffer.

Reply:
While I agree that we could improve it, I don't think there is an issue here currently. Only that it's a little hard to read.

Reply:
Correct. Anyway atm this improvement only works when permessage-deflate is disabled. What about `websocket.ping()` and `websocket.pong()`?

Reply:
> What about websocket.ping() and websocket.pong()?

We could do those as well. Didn't see them as a hot path.